### ISSUE TITLE
Easy Timeout

### ISSUE THREAD

**Issue by NoamWhy** [dotnet/csharplang#2539]
Try to guess what the following suggested syntax is designed for:

```csharp
void SomeFunction ()
{
	// Some initial logic ....

	timeout(10000)
	{
		// Some more logic ....
	}
	else
	{
		// Some alternative logic ....
	}

	// Some final logic ....
}
```
With this kind of syntax, implementing timeout functionality becomes trivial. 
Sure, timeout can be implemented today using various combinations of currently available multi-threading library functions, but none of these implementations is remotely as simple and clean as the code above.

P.S.
As a result of the discussion below, I am suggesting here the following better syntax:

```csharp
void SomeFunction ()
{
	// Some initial logic ....

	for (10000 milliseconds)
	{
		// do some more logic,
		// and if this block of code dose not finish within 10,000 milliseconds,
		// a TimeoutException will automatically be thrown
	}

	// Some final logic ....
}
```

**Comment by Joe4evr** [comment]
Why do you need new syntax for this? You can already do this with Tasks and CancellationTokens:
```cs
using (var cts = new CancellationTokenSource())
{
    cts.CancelAfter(10000);
    try
    {
        var result = await SomethingAsync(cts.Token);
        // Logic for success
    }
    catch (OperationCanceledException)
    {
        // Logic for timeout
    }
}
```
Of course, this requires the method itself to take in a CancellationToken, which is not always the case. For those cases, [you'll want to do what David Fowler recommends here](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#using-a-timeout).

**Comment by NoamWhy** [comment]
As I said:

> Sure, timeout can be implemented today using various combinations of currently available multi-threading library functions, but none of these implementations is remotely as simple and straightforward as the code above.

Here are few problems associated with the solution you offered above:

1.  It  requires moving the timed logic to a separate function, which would then prevent it from accessing local variables, and thus makes implementation harder.

2. It makes use of a try/catch mechanism, which may severely hurt performance. 

3. It requiters understanding of multi-threads programming, which is not trivial.

**Comment by Tom01098** [comment]
> It requires moving the timed logic to a separate function, which would then prevent it from accessing local variables, and thus makes implementation harder.

What do you mean by this @NoamWhy? In both the examples, the timeout login has been separated; in yours, it's been even more separated! I can't see how your mechanism accesses local variables or why this is a positive. Additionally, would this mechanism absorb Exceptions?

> It makes use of a try/catch mechanism, which may severely hurt performance.

Try-catch would still probably be implemented under the hood, like how a using statement implements one.

**Comment by NoamWhy** [comment]
> What do you mean by this @NoamWhy? In both the examples, the timeout login has been separated; in yours, it's been even more separated! I can't see how your mechanism accesses local variables or why this is a positive. 

Consider this example:

```csharp
void SomeFunction ()
{
	var URLs = new List<string>();

	// Build URLs list...

	var Results    = new List<string>();
	timeout(10000)
	{
		foreach (var URL in URLs)
		{
			Results.Add(GetDataFromURL(URL));
		}
	}
	else
	{
		// Some alternative logic ....
	}

	// Process results ....
}
```
 
> Try-catch would still probably be implemented under the hood, like how a using statement implements one.

Then this would not be the optimal implementation, if it hurts performance.

**Comment by Joe4evr** [comment]
> Then this would not be the optimal implementation, if it hurts performance.

Tough luck, because the CLR doesn't expose aborting a function outside of throwing.

**Comment by NoamWhy** [comment]
Also regarding this:

```csharp
using (var cts = new CancellationTokenSource())
{
    cts.CancelAfter(10000);
    try
    {
        var result = await SomethingAsync(cts.Token);
        // Logic for success
    }
    catch (OperationCanceledException)
    {
        // Logic for timeout
    }
}
```

What if you want to pass some other cancellation token you got to the SomethingAsync function?

**Comment by yaakov-h** [comment]
@NoamWhy Then you can use `CancellationTokenSource.CreateLinkedTokenSource`.

**Comment by NoamWhy** [comment]
> Tough luck, because the CLR doesn't expose aborting a function outside of throwing.

What's the total number of try/catch statements used in David Fowler's solution (including the ones hidden in all the using statements) ?

**Comment by Tom01098** [comment]
[Try-catch is not as slow as you think.](https://stackoverflow.com/questions/1308432/do-try-catch-blocks-hurt-performance-when-exceptions-are-not-thrown)
If you need this to be _incredibly_ fast then maybe try C++ or ASM.

Even so, the three try-catches in the solution would be needed implicitly anyway by the mechanism you propose.

**Comment by Tom01098** [comment]
To elaborate, I do like the idea - I still don't understand CancellationTokens in a good amount of detail! But I think the focus should be on more abstraction than performance.

**Comment by DavidArno** [comment]
> Try to guess what the following suggested syntax is designed for

My guess is that it is something to do with timeouts and that under some circumstances, the else-block is presumably run? Did I guess right?

Or to put it another way, asking us to guess what your proposal is, isn't very constructive. I literally am forced to guess what the code is supposed to do. So actually detailing the proposal, and explaining why additional syntax would be better than implementing it with methods, would be far more constructive.

**Comment by NoamWhy** [comment]
> My guess is that it is something to do with timeouts and that under some circumstances, the else-block is presumably run. Did I guess right?

You guessed correctly!

And this is precisely the point I was trying to make, i.e. this syntax is so simple and clean, that you can hardly misinterpret it. 

If you are still not sure why "additional syntax would be better than implementing it with methods", take a second look at the current alternative below, and see if it's as easy to guess what it's doing:

```csharp
void SomeFunc()
{
    try
    {
        var result = await TimeoutAfter(SomethingAsync( /*pass all needed local variables*/ ),
                                                           new TimeSpan(0,0,1000));
    }
    catch (OperationCanceledException)
    {
        // Logic for timeout
    }
}
public static async Task<T> TimeoutAfter<T>(this Task<T> task, TimeSpan timeout)
{
    using (var cts = new CancellationTokenSource())
    {
        var delayTask = Task.Delay(timeout, cts.Token);

        var resultTask = await Task.WhenAny(task, delayTask);
        if (resultTask == delayTask)
        {
            // Operation cancelled
            throw new OperationCanceledException();
        }
        else
        {
            // Cancel the timer task so that it does not fire
            cts.Cancel();
        }

        return await task;
    }
}
```

**Comment by theunrepentantgeek** [comment]
> And this is precisely the point I was trying to make, i.e. this syntax is so simple and clean, that you can hardly misinterpret it.

It's simple and clean *to you* because you created the idea, but there are a whole lot of questions that came to mind when I read this post ...

* What's the magic 10000 figure? 
* Is it some kind of flag? If so, what does it mean?
* If it's a duration, what are the units? Processor Ticks? Seconds? Jiffys? 
* What does the **else** clause do? 
* Normally an **else** separates two mutually exclusive execution paths - how is the decision made on whether to run the first branch or the second?

As @DavidArno wrote:
> asking us to guess what your proposal is, isn't very constructive. 

Every language features starts out at [-100 points](https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/). For any proposal to get traction, the proposal needs to clearly articulate not only why it's valuable, but why it is dramatically superior to what's already possible.

Another way to achieve a very similar result, would be to use a helper method like this:
```
void SomeFunction ()
{
	// Some initial logic ....

	RunWithTimeout(
        10.Seconds(),
        async () => {
            // Some more logic
        });

	// Some final logic ....
}
```

How is this so bad?

**Comment by spydacarnage** [comment]
While I agree with many of your points, the "magic 100000 in what units" is
on dodgy ground as the language already has that concept in other places.

**Comment by NoamWhy** [comment]
> there are a whole lot of questions that came to mind when I read this post ...
> 
> What's the magic 10000 figure?
> Is it some kind of flag? If so, what does it mean?
> If it's a duration, what are the units? Processor Ticks? Seconds? Jiffys?
> What does the else clause do?
> Normally an else separates two mutually exclusive execution paths - how is the decision made on whether to run the first branch or the second?

Here is a better syntax, if you will:

```csharp
void SomeFunction ()
{
	// Some initial logic ....

	for (10000 milliseconds) try
	{
		// Some more logic ....
	}
	catch(TimeoutException)
	{
		// Some alternative logic ....
	}

	// Some final logic ....
}
```

**Comment by johnkellyoxford** [comment]
> It makes use of a try/catch mechanism, which may severely hurt performance.

No. Without a throw, the cost of a `try` is trivial. 

The thing is, the alternative to this feature is A LOT simpler - just don't add it. Every langauge feature needs to be very useful to be added. This is just a little syntactic sugar for a pattern which isn't massively common.

**Comment by NoamWhy** [comment]
> This is just a little syntactic sugar for a pattern which isn't massively common

Off course it's not massively common, because it's a pain to implement!

As a matter of best practice, virtually any call to external resources should be protected by timeout, if you want your code to be responsive and not get stuck. 

This syntax would not only make our life easier, but would actually make us write better code.

**Comment by DavidArno** [comment]
> take a second look at the current alternative below, and see if it's as easy to guess what it's doing

I don't need to guess there: I can see precisely what it's doing. And in the process, I now understand what your proposal is doing. I read you proposal as after some timeout (and I was confused as to what was being measured for a timeout), run `// Some more logic ....` and if it didn't time out, then do the `else` part. But you are proposing (I think) that if `// Some more logic ....` doesn't complete in the timeout period, then the `else` part is run.

So you are (I think) proposing that:
```cs
timeout(time)
{
    // some processes that must finish within "time" milliseconds
}
else
{
    // some code that's run if the previous block didn't complete
}
```
be used instead of:
```cs
try
{
    Task.Run (() => 
    {
        // some processes that must finish within "time" milliseconds
    }).TimeoutAfter(new TimeSpan(0,0,time)).Wait();
}
catch (OperationCanceledException)
{
    // some code that's run if the previous block didn't complete
}
```
Is that correct?

**Comment by NoamWhy** [comment]
> So you are (I think) proposing that:
> ```csharp
> timeout(time)
> {
>     // some processes that must finish within "time" milliseconds
> }
> else
> {
>     // some code that's run if the previous block didn't complete
> }
>  ```
> be used instead of:
>  ```csharp
> try
> {
>     Task.Run (() => 
>     {
>         // some processes that must finish within "time" milliseconds
>     }).TimeoutAfter(new TimeSpan(0,0,time)).Wait();
> }
> catch (OperationCanceledException)
> {
>     // some code that's run if the previous block didn't complete
> }
> ```
> Is that correct?

Yes, and an even better syntax would be:

```csharp
void SomeFunction ()
{
	// Some initial logic ....

	for (10000 milliseconds) try
	{
		// some processes that must finish within 10,000 milliseconds
	}
	catch(TimeoutException)
	{
		// some code that's run if the previous block didn't complete within 10,000 milliseconds
	}

	// Some final logic ....
}
```

**Comment by NoamWhy** [comment]
Buy the way, I believe the following code would not work:

```csharp
try
{
    Task.Run (() => 
    {
        // some processes that must finish within "time" milliseconds
    }).TimeoutAfter(new TimeSpan(0,0,time)).Wait();
}
catch (OperationCanceledException)
{
    // some code that's run if the previous block didn't complete
}
```
This is because the TimeoutAfter function runs in a task of it's own, and hence any exception it throws would not propagate up.

**Comment by NoamWhy** [comment]
Here is yet a simpler syntax.

```csharp
void SomeFunction ()
{
	// Some initial logic ....

	for (10000 milliseconds)
	{
		// do some more logic,
		// and if this block of code dose not finish within 10,000 milliseconds,
		// a TimeoutException will automatically be thrown
	}

	// Some final logic ....
}
```

**Comment by HaloFour** [comment]
Syntax aside, how exactly do you propose that this actually be implemented?  What enforces the timeout and what causes the exception to be thrown?  The current thread is already busy doing "do some more logic", which could certainly include the invocation of other functions or blocking operations.

**Comment by NoamWhy** [comment]
> Syntax aside, how exactly do you propose that this actually be implemented? What enforces the timeout and what causes the exception to be thrown? The current thread is already busy doing "do some more logic", which could certainly include the invocation of other functions or blocking operations.

No, under the hood the current thread should create another thread that performs "some more logic", while waiting for a timer.

**Comment by HaloFour** [comment]
> No, under the hood the current thread should create another thread that performs "some more logic", while waiting for a timer.

That would be quite expensive and opens up threading and synchronization issues that wouldn't otherwise exist.  Such behavior would be much better suited to a library method using CancellationTokens and Tasks.

**Comment by HaloFour** [comment]
Not to mention, you decried closure/delegate based solutions above, yet that's exactly what the compiler must do if it's going to spin up a separate thread to execute that block of logic.

**Comment by NoamWhy** [comment]
> That would be quite expensive and opens up threading and synchronization issues that wouldn't otherwise exist. Such behavior would be much better suited to a library method using CancellationTokens and Tasks. Not to mention, you decried closure/delegate based solutions above, yet that's exactly what the compiler must do if it's going to spin up a separate thread to execute that block of logic.

I'll take your expert word on this. I'm not fixed on any particular implementation. whatever will work is fine with me.

**Comment by HaloFour** [comment]
There's probably a dozen or so ways something like this could be implemented, each having different tradeoffs.

Any implementation involving threading would increase the complexity and cost dramatically as you immediately have to deal with closures and synchronization.  There's also the question of scheduling.  Creating a new thread adds overhead and is quite wasteful if the parent thread is just going to block.  Using a thread pool decreases the overhead of thread creation but could lead to other issues around queuing work, pool exhaustion, etc..  The scheduler is something that the developer would ultimately want to decide, not the language.

If the timed task itself is intended to be cancelled if it takes too long you open a second big can of worms.  The only really safe way to do that is cooperatively, generally through something like a CancellationToken.  Anything that would otherwise abort/interrupt the thread could leave the application in an intermediate state.  Interrupting the thread would be the safer of the options but only automatically works for functions that put the thread into a `WaitSleepJoin` state such as sleeping or blocking on I/O.  CPU bound tasks will merrily continue to chug along.  Thread aborting is quite unsafe and isn't even implemented in .NET Core.

Probably the least expensive way to accomplish this would be to create a CancellationToken and then use on the current thread to determine if the token has been cancelled due to a timeout.  But that only works for the code that you've written to check the token.  Any other function that you call that doesn't accept and respect the token would be unaware of this timeout and defeat the feature entirely.

Given that such a feature is loaded with policy, my opinion is that it doesn't belong in the language.  This is something that could be accomplished relatively easily with a helper function that accepts a delegate for the task, a cancellation token and a scheduler, like [`TaskFactory.StartNew`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskfactory.startnew?view=netframework-4.8#System_Threading_Tasks_TaskFactory_StartNew__1_System_Func_System_Object___0__System_Object_System_Threading_CancellationToken_System_Threading_Tasks_TaskCreationOptions_System_Threading_Tasks_TaskScheduler_).

**Comment by NoamWhy** [comment]
> There's probably a dozen or so ways something like this could be implemented, each having different tradeoffs...

I am not sure I fully understand your opinion. 
You begin by explaining how challenging implementing timeout is - which then would completely justify such a built in feature, in order to make life easier for programmers. 
However, your final statement says "this is something that could be accomplished relatively easily..." - so here is your easy way to implement this suggested feature. 

Which way is it?

**Comment by HaloFour** [comment]
@NoamWhy 

**TL;DR**:  This feature doesn't belong in the language because there are too many possible implementations each loaded with policy, behavioral caveats and performance characteristics to railroad the developer into one single version.  You can't just wrap some arbitrary block of code in brackets and expect anything like this to work.

For specific versions of the implementation, you can accomplish something like it with helper functions like the one I described.  But you are very explicitly opting into closures, delegates, threading, task schedulers and cancellation concerns.  Your task needs to be written to be cancellable, as does everything that it calls.  If that's fine for your use case, call the helper function.

**Comment by CyrusNajmabadi** [comment]
> which then would completely justify such a built in feature, in order to make life easier for programmers.

It woudln't though, because the right sort of approach is domain specific.

**Comment by NoamWhy** [comment]
> It wouldn't though, because the right sort of approach is domain specific.

So what you're basically saying is - since there are several different ways to implement this, the language  will offer none, not even the most common, and we will leave it all for programmers to navigate this complex task of figuring out how to best implement timeout functionality.

**Comment by CyrusNajmabadi** [comment]
> So what you're basically saying is - since there are several different ways to implement this, the language will offer none, not even the most common,

What is "the most common".

>  and we will leave it all for programmers to navigate this complex task of figuring out how to best implement timeout functionality.

Correct.  Because it's domain specific.  And the language doesn't want to be opinionated in a case like this.  That's why we have libraries and APIs.  So that people can pick between the myriad pros and cons based on their specific needs.

**Comment by NoamWhy** [comment]
> Correct. Because it's domain specific. And the language doesn't want to be opinionated in a case like this. That's why we have libraries and APIs. So that people can pick between the myriad pros and cons based on their specific needs.

Can you give me your implementation for this?

```csharp
void SomeFunction ()
{
	// Some initial logic ....

	for (10000 milliseconds)
	{
		// do some more logic,
		// and if this block of code dose not finish within 10,000 milliseconds,
		// a TimeoutException will automatically be thrown
	}

	// Some final logic ....
}
```

**Comment by HaloFour** [comment]
@NoamWhy 

> Can you give me your implementation for this?

No, I can't.  Not without knowing exactly what "do some more logic" is.  There's quite literally no way to have a default way to implement this.  Nor should there be.

**Comment by HaloFour** [comment]
@NoamWhy 

> and we will leave it all for programmers to navigate this complex task of figuring out how to best implement timeout functionality.

We leave it to programmers to find helper functions, just like we do the other 99.99% of the time.  If the helper functions that exist don't behave in the way that the developer wants, they can write another.  If this was a language feature the developer would be otherwise stuck between whatever opinions the language design team chose, or having to find other helper functions anyway.  Not everything can/should be a language feature.

**Comment by NoamWhy** [comment]
> No, I can't. Not without knowing exactly what "do some more logic" is. There's quite literally no way to have a default way to implement this. Nor should there be.

Ok, suppose  "do some more logic" = "call a bunch of 3rd party functions you have no control of"

**Comment by HaloFour** [comment]
@NoamWhy 

> Ok, suppose "do some more logic" = "call a bunch of 3rd party functions you have no control of"

There is no good answer then.  You can start that logic in a separate Task and then wait on that task for a given amount of time, but there's no good way to cancel that "do some more logic".  You can try to interrupt the thread, but that only applies to specific kinds of workloads.

**Comment by CyrusNajmabadi** [comment]
> Can you give me your implementation for this?

It would very much depend on that overall concurrency approach being taken .  What general threading/concurrency approach do you want to take here?

**Comment by NoamWhy** [comment]
> It would very much depend on that overall concurrency approach being taken . What general threading/concurrency library or approach do you want to take here?

Let's say Tasks.

**Comment by CyrusNajmabadi** [comment]
Tasks would be very tricky.  Because you'd need a way to encode cancellability, as well as determining what the sync-context rules would be.

**Comment by NoamWhy** [comment]
> Tasks would be very tricky. Because you'd need a way to encode cancellability, as well as determining what the sync-context rules would be.

Then let's start with the easiest concurrency model of your choice.

**Comment by NoamWhy** [comment]
And by the way, there is no requirement to perform any cancellation. The only thing that is required is to raise a TimeoutException. I believe this would makes implementation easier.

**Comment by CyrusNajmabadi** [comment]
> Then let's start with the easiest concurrency model of your choice.

Er... i'd never describe concurrency as easy.  It's an extremely tricky area.  Take roslyn itself (the C# compiler and IDE).  There are multiple different concurrency models used (TPL, JTF, Parallel-Extensions, and just plain threading), at many different layers.  Even in individual layers, we use a variety of forms depending on what we care most about.

**Comment by jnm2** [comment]
There is no safe default way to implement this in the language. Abandoning or aborting a thread is a very bad idea in general. Cooperative cancellation is great, but it means that the code in the `timeout` block must use a cancellation token.

**Comment by NoamWhy** [comment]
> Er... i'd never describe concurrency as easy. It's an extremely tricky area. Take roslyn itself (the C# compiler and IDE). There are multiple different concurrency models used (TPL, JTF, Parallel-Extensions, and just plain threading), at many different layers. Even in individual layers, we use a variety of forms depending on what we care most about.

I think that cancellation is the problem. But as I said:

> there is no requirement to perform any cancellation. The only thing that is required is to raise a TimeoutException. I believe this would makes implementation easier.

We can leave it for the programmer to decide what to do to when such an exception is raised.

**Comment by HaloFour** [comment]
@NoamWhy 

Raise the `TimeoutException` where?

**Comment by CyrusNajmabadi** [comment]
> there is no requirement to perform any cancellation. The only thing that is required is to raise a TimeoutException. I believe this would make implementation easier.

How would you do this?  Can you give a full example, including where you think the timeoutexception would occur?

**Comment by CyrusNajmabadi** [comment]
Based on previous discussions, let me jump back and say: thanks for proposing this issue.  These are really valuable for discussion and to explore the problem space.  The feedback you're getting is not meant to be negative or to push you away from exploring this space.  It's intended to help you (and everyone else) consider this area in full to understand the complexity and challenges here.

Thanks again!

**Comment by jnm2** [comment]
@Joe4evr It could also be:


```cs
using (var cts = new CancellationTokenSource(delay: TimeSpan.FromSeconds(10)))
{
    try
    {
        var result = await SomethingAsync(cts.Token);
        // Logic for success
    }
    catch (OperationCanceledException)
    {
        // Logic for timeout
    }
}
```

**Comment by NoamWhy** [comment]
> @Joe4evr It could also be:
> 
> ```csharp
> using (var cts = new CancellationTokenSource(delay: TimeSpan.FromSeconds(10)))
> {
>     try
>     {
>         var result = await SomethingAsync(cts.Token);
>         // Logic for success
>     }
>     catch (OperationCanceledException)
>     {
>         // Logic for timeout
>     }
> }
> ```

This would not work if SomethingAsync is stuck and doesn't handle the token, or does not accept one in the first place.

**Comment by jnm2** [comment]
@NoamWhy Cooperative cancellation is in general the only safe thing to do. In that scenario, the best thing to do is fix SomethingAsync; the alternatives are bad.

**Comment by NoamWhy** [comment]
Here is something I have in mind, though I'm not sure if it would work...

void SomeFunction ()
{
    using (var timer = new Timer(delegate (object x) { throw new TimeoutException(); }, null, 10000, 0))
    {
        // Do some timed logic
    }
}

**Comment by jnm2** [comment]
For example, before JSON.NET added cancellation tokens to its async reader methods, one option was to buffer the stream to a memory stream (with the ability to cancel) and then parse the buffered stream synchronously without the ability to cancel. Another riskier option might be to register a cancellation token callback that closes the stream and then handle the exception `when (cancellationToken.IsCancellationRequested)` with `cancellationToken.Throw...();`

These aren't things that the compiler could know to do, but they are some of the best options.

Throwing the exception in one thread won't affect another thread; it'll just mess with the thread pool thread which the timer scheduled the callback on.

**Comment by NoamWhy** [comment]
So here is another way

```csharp
void SomeFunction ()
{

   var TimerTask = Task.Delay(1000).Start();

   if(TimerTask ==  Task.WaitAny 
        (
            TimerTask
            ,Task.Run
             (  () => 
                 {
                   // Do some timed logic
                 }
              )
       )
   ) throw new TimeoutException();
}
```

**Comment by CyrusNajmabadi** [comment]
> Here is something I have in mind, though I'm not sure if it would work...

This would likely not at all do what you want :-/     This will just throw that exception on a random threadpool thread.  By default, you'll just get an unhandled exception error that will tear down your app.

**Comment by CyrusNajmabadi** [comment]
> So here is another way

Right, but major problems here are that you're now using the threadpool.  Having normal C# code suddenly jumping threads (especially to the threadpool) would likely be extremely surprising for people.  it would also be a non-starter in many (most?) domains.

**Comment by NoamWhy** [comment]
> Also, how would you make it so that things like "return" would work? 

```csharp
void SomeFunction ()
{

   var TimerTask = Task.Delay(1000).Start();
   var QuitFun = true;
   if(TimerTask ==  Task.WaitAny 
        (
            TimerTask
            ,Task.Run
             (  () => 
                 {
                   // Do some timed logic

                    QuitFun = false;
                 }
              )
       )
   ) throw new TimeoutException();
   else if (QuitFun) return;
}
```

**Comment by CyrusNajmabadi** [comment]
That won't actually work.  You need to await Task.WhenAny.  The 'Task' you get back from it isn't TimerTask (or the Task from Task.Run) it's a *new* task that *when awaited* will give you one of those tasks back.  

I would highly recommend trying to make a working example of how the *translated* code would even look like.  :)

**Comment by NoamWhy** [comment]
> That won't actually work. You need to await Task.WhenAny. The 'Task' you get back from it isn't TimerTask (or the Task from Task.Run) it's a new task that when awaited will give you one of those tasks back.
> 
> I would highly recommend trying to make a working example of how the translated code would even look like. :)

Not true, since I am using Task.WaitAny and not Task.WhenAny.
Actually, it is much simpler, since WaitAny returns an index, not a task:

```csharp
void SomeFunction ()
{

   var QuitFun = true;

   if(0 ==  Task.WaitAny 
        (
            Task.Delay(1000).Start();
            ,Task.Run
             (  () => 
                 {
                   // Do some timed logic

                    QuitFun = false;
                 }
              )
       )
   ) throw new TimeoutException();
   else if (QuitFun) return;
}
```

**Comment by HaloFour** [comment]
The fact that we're reaching for Tasks or any threading mechanism I think is reason enough that this can't be a simple language feature.  There are just way too many gotchas here that a simple language feature would hide away from the developer.  For example, the fact that the calling thread would block is a very common vector for deadlocks, and why mixing async/sync is generally considered a Bad Idea.

**Comment by NoamWhy** [comment]
> There are just way too many gotchas here that a simple language feature would hide away from the developer

That's precisely the whole purpose of high level languages.

> For example, the fact that the calling thread would block is a very common vector for deadlocks

I don't see the problem, can you describe such a scenario?

> mixing async/sync is generally considered a Bad Idea.

No such mix is suggested. For sync functions we will use WaitAny, and for async we will use WhenAny and await.

**Comment by johnkellyoxford** [comment]
If you really want it, why not propose it as a simple all in one function in dotnet/corefx rather than as a language feature?

**Comment by HaloFour** [comment]
@NoamWhy 

> That's precisely the whole purpose of high level languages.

It's not the purpose of high level languages to take everything that can be done in helper functions and turn it into language features.

> For sync functions we will use WaitAny

Which is blocking (sync) an async operation (the Task).

One example is if the calling method holds a `lock` on some reference.  Then within the `timeout` block you also try to take a `lock` on that resource.  If it was on the same thread that would be fine since locks are reentrant, but since you've now moved that logic to a separate thread that will block indefinitely, or at least until the timeout expires.

Another example is UI interaction.  Of the calling method is on the UI thread then the code within `timeout` suddenly can't interact directly with UI elements.  The way to do that safely is via `Control.Invoke` which dispatches back to the UI thread.  But of course the UI thread is blocking waiting on the timeout code.  That will also block indefinitely until the timeout expires.

**Comment by NoamWhy** [comment]
@HaloFour

Fair enough, so programmers should be aware of all these issues, but this is not an excuse to abandon altogether any attempt to simplify timeout implementation.

**Comment by johnkellyoxford** [comment]
Yes, so make a method to do it for you, and suggest it is added to the standard library. Why is a language feature needed?

**Comment by NoamWhy** [comment]
> Yes, so make a method to do it for you, and suggest it is added to the standard library. Why is a language feature needed?

Why do we need foreach (...) if we can use for (...) and while(Enum.MoveNext())?
Why do we need await and async, if we can use worker threads?
Why do we need var x="some text", if we can write string x = "some text"? 

The answer to all these question is - efficiency, clarity, and ease.

**Comment by HaloFour** [comment]
@NoamWhy 

> Fair enough, so programmers should be aware of all these issues, but this is not an excuse to abandon altogether any attempt to simplify timeout implementation.

It is a big reason for not trying to embed it into the language with syntax that belies all of the concerns you need to be deeply aware of in order to use it.

> The answer to all these question is - efficiency, clarity, and ease.

None of them have the problems that this feature has.

**Comment by NoamWhy** [comment]
> None of them have the problems that this feature has.

Simply inform programmers that the timed code is executed in a separate thread/task, and should be treated like any other such code. Done!

**Comment by johnkellyoxford** [comment]
> Why do we need foreach (...) if we can use for (...) and while(Enum.MoveNext())?

People use them significantly more than a timeout feature. And seemingly everyone except you here agrees on that

**Comment by NoamWhy** [comment]
> People use them significantly more than a timeout feature. And seemingly everyone except you here agrees on that

As I wrote above:

> Off course timeout is not massively common, because it's a pain to implement!
> 
> As a matter of best practice, virtually any call to external resources should be protected by timeout, if you want your code to be responsive and not get stuck.
> 
> This syntax would not only make our life easier, but would actually make us write better code.
>

**Comment by Tom01098** [comment]
Unfortunately if it's not common, it's very unlikely to be implemented - the C# team have a limited amount of time to work on things with broad application like NRTs.
As mentioned above, if you wrote a helper method for CoreFX to ease the pain of implementation, then everyone's a winner.

**Comment by FiniteReality** [comment]
Timeouts in general are much better suited to a programming library rather than the language itself. Remember, C# doesn't specify the threading model or whatever for a lot of code - especially async code - that's done entirely in .NET, using types like TaskScheduler and ThreadPool.

Adding support for timeouts directly in the language would mean C# *would* have to start specifying threading models, concurrency approaches, etc. and that means that it would take a lot of time to even think about getting right. And then what happens if for a specific person they get those options wrong? They'd be fighting against the language to get their specific task done, and likely making lots of mistakes along the way.

To me, this honestly feels like a Pit Of Despair solution to a problem that can easily be solved in library code. For example, you could have a method:
```cs
Task<TResult> CancelAfter<TResult>(TimeSpan timeout, Func<CancellationToken, Task<TResult>> callback) { throw null; }
```
And this one method would look, honestly, just as nice as your syntax:
```cs
async Task MyMethodAsync()
{
    await CancelAfter(TimeSpan.FromSeconds(5),
        async (cancelToken) =>
        {
            // code here
            cancelToken.ThrowIfCancellationRequested();
        });
}
```
